include/envoy/thread/thread.h

```diff
@@ -2,40 +2,115 @@
 
 #include "envoy/common/pure.h"
 
+#include "absl/base/thread_annotations.h"
+
 namespace Envoy {
 namespace Thread {
 
 /**
- * Like the C++11 "basic lockable concept" but a concrete interface vs. a template.
+ * Like the C++11 "basic lockable concept" but a pure virtual interface vs. a template, and
+ * with thread annotations.
  */
-class BasicLockable {
+class LOCKABLE BasicLockable {
 public:
   virtual ~BasicLockable() {}
 
-  virtual void lock() PURE;
-  virtual bool try_lock() PURE;
-  virtual void unlock() PURE;
+  virtual void lock() EXCLUSIVE_LOCK_FUNCTION() PURE;
+  virtual bool tryLock() EXCLUSIVE_TRYLOCK_FUNCTION(true) PURE;
+  virtual void unlock() UNLOCK_FUNCTION() PURE;
 };
 
 /**
  * A lock guard that deals with an optional lock.
  */
-template <class T> class OptionalLockGuard {
+class SCOPED_LOCKABLE OptionalLockGuard {
```

Comment:
IMO it was not useful to make this templated, as we've just defined the interface that the passed-in lock must implement, so it seemed like a redundant mechanism.

I think the only practical value was to pass in a std::mutex* (thus depending on the try_lock API spelling above) but given that std::mutex has no annotations, that's kind of a dead end.  So I think it makes sense to simplify by removing the templating here.

Reply:
Yes makes sense. Again I think I did this to work with std:: stuff so fine to change.